BackendTaskPilot/Program.cs

```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
+
+// REST API to get exchange rates based on input currencies
+app.MapPost("/", service.GetExchangeRates);
+
+app.Run();
+
+public struct CnbRatesResponse
+{
+    [JsonPropertyName("rates")] public List<Rate> Rates { get; set; }
+}
+
+public struct Rate
+{
+    [JsonPropertyName("currencyCode")] public string CurrencyCode { get; set; }
+    [JsonPropertyName("rate")] public float RateValue { get; set; }
+}
+
+internal class ExchangeRateService
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
+    {
+        HttpClient client = new HttpClient();
```

Comment:
Creating a new `HttpClient` for every request to our service is not a great idea, as creating a lot of `HttpClient`s like this can lead to resource issues under heavy load.

We want to centralize creation and configuration to make sure it only happens when necessary. Luckily, ASP.NET gives us an easy way to do this. At the top of `Program.cs`, we ask the application `builder` to create and configure the `HttpClient` for us:

```
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddHttpClient(); // <---------------- added line
```

This sets up a so called *client factory* which we can then use give us correctly configured clients for us whenever we need them. It uses an internal logic to save on resource usage. Through a process known as **dependency injection**, often abbreviated DI, we can get a hold of our factory where we want it by specifying it as a parameter to the method used to handle the API request. 

Note the added parameter at the end:

```
public async IEnumerable<Rate> GetExchangeRates(List<string> currencies, IHttpClientFactory clientFactory)
```

We can the use it inside our method to request a client for our usage:

`var client = clientFactory.CreateClient(); // used to be var client = new HttpClient()`

While this might seem overly complex compared to `new HttpClient()` there are a lot of benefits that might not be obvious at first. It's a good habit to learn early in order to avoid nasty surprises down the line.